[docs] Some changes to gtk intro

sophie-h opened the pull request.

bilelmoussaoui commented on gtk/src/lib.rs:
```diff
@@ -16,71 +16,71 @@
+//!         .text("Hello, World!")
+//!         .buttons(ButtonsType::Close)
+//!         .build()
+//!         .run();
```
Should use show here

sdroege replied:
Should maybe be something more useful here like the Hello World on https://gtk.org (which is about the same number of lines...)

sophie-h replied:
The example would not work with show because there is no main loop. If you think it's worth it, I can restructure the examples. For gtk4 I would just give one example and mostly link to the book for the rest.

sophie-h replied:
I will just drop this example, there is a proper example.

bilelmoussaoui replied:
> The example would not work with show because there is no main loop

We should replace that with something that the user can copy/paste and have something working.

sdroege replied:
Yeah dropping the example and just keeping the one below makes most sense, I guess (and maybe replacing the example below with the gtk.org one).

sdroege commented on gtk/src/lib.rs:
```diff
@@ -129,9 +129,9 @@
 //! users will have it installed.
 //!
-//! ## Lgpl-docs
+//! ### Lgpl-docs
```
```suggestion
//! ### LGPL docs
```

sdroege reviewed (commented):
Seems good to me otherwise :)

GuillaumeGomez commented on gtk/src/lib.rs:
```diff
@@ -16,71 +16,71 @@
+//! - [GTK documentation](https://www.gtk.org/docs/)
 //!
-//! # Hello World
+//! ## “Hello, World!” example program
```
There is no need for the "fancy" (don't know what they're called) double-quote, rustdoc is changing them automatically (as you can see [here](https://doc.rust-lang.org/nightly/std/primitive.never.html)).

GuillaumeGomez commented on gtk/src/lib.rs:
```diff
@@ -16,71 +16,71 @@
 //! See also:
 //!
-//! - [Gtk-rs documentation overview](https://gtk-rs.org/docs-src/)
+//! - [gtk-rs documentation overview](https://gtk-rs.org/docs-src/)
```
In most places, we use `Gtk-rs` (for example <https://opencollective.com/gtk-rs> and <https://gtk-rs.org>). So for now, I think it'd be better to keep as is until we make a decision about it.

sdroege replied:
Should probably be `gtk-rs` or `GTK-rs`, but let's handle that separately I guess...

GuillaumeGomez replied:
Agreed (for context: `gtk-rs` refer to the repository whereas `Gtk-rs` currently somewhat refers to the organization. It's very unclear).

GuillaumeGomez commented on gtk/src/lib.rs:
```diff
@@ -3,10 +3,10 @@
-//! This library contains safe Rust bindings for [GTK+ 3](http://www.gtk.org), a
-//! multi-platform GUI toolkit. It's a part of [Gtk-rs](http://gtk-rs.org/).
+//! This library contains safe Rust bindings for [GTK 3](http://www.gtk.org), a
+//! multi-platform GUI toolkit. It's a part of [gtk-rs](http://gtk-rs.org/).
```
Same as below.

GuillaumeGomez commented on gtk/src/lib.rs:
```diff
@@ -129,9 +129,9 @@
+//! ### Lgpl-docs
 //!
-//! The Gtk-rs crates come with API docs missing because of licensing
+//! The gtk-rs crates come with API docs missing because of licensing
```
Same as above.

sdroege commented on gtk/src/lib.rs:
```diff
@@ -20,67 +20,81 @@
-//! do it once and there is no 'finalize'.
+//! GTK needs to be initialized before use by calling [function@init] or by
+//! creating an [struct@Application]. You only need to do it once and there
+//! is no 'finalize'.
```
> You only need to do it once and there is no 'finalize'.

I don't know what this means (even before your changes) :) @GuillaumeGomez?

GuillaumeGomez replied:
I meant that you don't need to free it. (Yep, can definitely be improved haha)

sdroege replied:
The sentence is gone now as it makes not much sense :) Memory is managed automatically in the bindings everywhere.

sophie-h replied:
Yes! This is rust here! :laughing:

sdroege commented on gtk/src/lib.rs:
```diff
@@ -20,67 +20,81 @@
+//! - [GTK documentation](https://www.gtk.org/docs/)
 //!
-//! # Hello World
+//! ## "Hello, World!" example program
```
I think this can be merged with the next section, the code is the same anyway

sdroege commented on gtk/src/lib.rs:
```diff
@@ -102,20 +112,20 @@
 //! such cases will be documented individually but they're not yet.
 //!
-//! **A panic in a closure will abort the process.**
+//! **A panic in a closure will abort the main thread.**
```
That's not true. It will actually kill the whole process

sophie-h replied:
I think you can spawn a thread an do the init there? But I can still revert it.

sophie-h replied:
I was thinking about specifying "in a closure" a bit more. The sentence as it is seems a bit unclear.

sdroege replied:
Any panic in a signal handler closure or any other closure passed to bindings functions will directly kill the whole process. The panic would unwind until an `extern "C"` function, and there it would then call `abort()`.

With current stable Rust it's actually UB (but in almost all cases does exactly that already) but next (or next-next?) release it's guaranteed to do that.

sdroege replied:
But yes, making the "in a closure" part a bit more clear is a good idea :)

sophie-h commented:
- Removed all the bold formatting because it's not helping imo. I some cases it might make things worse.
- I have removed the statement about missing bindings because I think gtk-rs is quite complete?
- I think it would be good to have an explicit statement about gtk-dev being an external requirement for gtk-rs to work.
- Looks like we can almost copy paste this for gtk4. Just at a prominent mention of the book.

sdroege commented:
Seems all good to me like this. For `gtk-dev`, no opinion here :)

sophie-h commented:
> For `gtk-dev`, no opinion here :)

Let's merge this first. Potentially mentioning distro packages etc. is tedious. (I still think it's important though.)

sdroege commented:
> Let's merge this first. Potentially mentioning distro packages etc. is tedious. (I still think it's important though.)

I agree. Also why is `gtk-dev` not required? That's where the pkg-config files are, or not?

sophie-h commented:
> Also why is `gtk-dev` not required?

I'm not following. Who said it's not required?

sdroege commented:
Nobody, I somehow read a "not" in your initial statement that doesn't actually exist. Sorry for the confusion :)

sophie-h commented:
Can this and #435 be merged?

GuillaumeGomez commented:
Thanks!
